packages/runtime/container-runtime/src/opLifecycle/opGroupingManager.ts

```diff
@@ -95,11 +95,14 @@ export class OpGroupingManager {
 		return groupedBatch;
 	}
 
+	private lastSeenSeqNum = -1;
 	public ungroupOp(op: ISequencedDocumentMessage): ISequencedDocumentMessage[] {
 		let fakeCsn = 1;
 		if (!isGroupContents(op.contents)) {
 			// Align the worlds of what clientSequenceNumber represents when grouped batching is enabled
-			if (this.config.groupedBatchingEnabled) {
+			// If lastSeenSeqNum is a match, we know we already processed this op
```

Comment:
Can you clarify why this works and when it may stop working - if for example we change the logic inside the remote message processor? Seems like this is coupled to the way we can `ungroupOp` in a way that's hard to reason about.

To me, reading this comment tells me that if the last ungrouped op has the same sequenceNumber as the one we're processing right now, we've called `ungroupOp` twice consecutively for the same ungrouped op (in the meantime, we may have `ungroupOp` calls for actually grouped ops).

Reply:
This works because it's only assigning `clientSeqNum` 1 when we see a new `seqNum`. The times where it will match is when processing messages in an ungrouped batch after index 1.
A way it may break is if the `RemoteMessageProcessor` calls `ungroupOp` in a random order.
 
Maybe it just makes sense to remove this code altogether as it is more prone to break, is hard to read/decipher, and is stateful. `clientSeqNum` will mean 2 different things depending on a batch being grouped or not, but that likely isn't an issue at all.